### DISCUSSION TITLE
3.2 upgrade NETBOX_DELETE_LEGACY_DATA?

### DISCUSSION THREAD

**Discussion by jemery27** [netbox-community/netbox#9086] [Q&A] [answered]
Trying to run the 3.2 upgrade and getting error about ASN data. This isn;t something that would matter to us, but how do I tell the upgrade to proceed and delete the old data? The error says to "set the NETBOX_DELETE_LEGACY_DATA environment variable" - sorry but where does that need to be set?

Thank you,
Jason

**Comment by jeremystretch** [comment]
The environment variable just needs to be defined when the upgrade script is run. You can, for example, do:

```
$ export NETBOX_DELETE_LEGACY_DATA=1
$ sudo upgrade.sh
```

**Reply by jemery27** [reply]
Ok thanks I tried setting it to True. But actually that still didn't seem to work. I do see it set in printenv but get the same error on the upgrade:

  Apply all migrations: admin, auth, circuits, contenttypes, dcim, django_rq, extras, ipam, sessions, social_django, taggit, tenancy, users, virtualization, wireless
Running migrations:
  Applying dcim.0145_site_remove_deprecated_fields...Traceback (most recent call last):
  File "netbox/manage.py", line 10, in <module>
    execute_from_command_line(sys.argv)
  File "/opt/netbox/venv/lib/python3.8/site-packages/django/core/management/__init__.py", line 446, in execute_from_command_line
    utility.execute()
  File "/opt/netbox/venv/lib/python3.8/site-packages/django/core/management/__init__.py", line 440, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/opt/netbox/venv/lib/python3.8/site-packages/django/core/management/base.py", line 414, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/opt/netbox/venv/lib/python3.8/site-packages/django/core/management/base.py", line 460, in execute
    output = self.handle(*args, **options)
  File "/opt/netbox/venv/lib/python3.8/site-packages/django/core/management/base.py", line 98, in wrapped
    res = handle_func(*args, **kwargs)
  File "/opt/netbox/venv/lib/python3.8/site-packages/django/core/management/commands/migrate.py", line 290, in handle
    post_migrate_state = executor.migrate(
  File "/opt/netbox/venv/lib/python3.8/site-packages/django/db/migrations/executor.py", line 131, in migrate
    state = self._migrate_all_forwards(
  File "/opt/netbox/venv/lib/python3.8/site-packages/django/db/migrations/executor.py", line 163, in _migrate_all_forwards
    state = self.apply_migration(
  File "/opt/netbox/venv/lib/python3.8/site-packages/django/db/migrations/executor.py", line 248, in apply_migration
    state = migration.apply(state, schema_editor)
  File "/opt/netbox/venv/lib/python3.8/site-packages/django/db/migrations/migration.py", line 131, in apply
    operation.database_forwards(
  File "/opt/netbox/venv/lib/python3.8/site-packages/django/db/migrations/operations/special.py", line 193, in database_forwards
    self.code(from_state.apps, schema_editor)
  File "/opt/netbox/netbox/dcim/migrations/0145_site_remove_deprecated_fields.py", line 15, in check_legacy_data
    raise DataError(
django.db.utils.DataError: Unable to proceed with deleting asn field from Site model: Found 4 sites with legacy ASN data. Please ensure all legacy site ASN data has been migrated to ASN objects before proceeding. Or, set the NETBOX_DELETE_LEGACY_DATA environment variable to bypass this safeguard and delete all legacy site ASN data.


ubuntu@ip-10-26-255-26:/opt/netbox$ echo $NETBOX_DELETE_LEGACY_DATA 
1
ubuntu@ip-10-26-255-26:/opt/netbox$

**Comment by jeremystretch** [comment, marked as answer]
Try:

```
$ sudo NETBOX_DELETE_LEGACY_DATA=1 upgrade.sh
```

Seems to be a bit platform-dependent.

**Reply by kkthxbye-code** [reply]
You normally need the -E option to pass exported environment variables to sudo. But setting it in the command should probably work too.

**Reply by jemery27** [reply]
That worked for me - thank you!

**Reply by github6043** [reply]
Hi jemery27, have the same problem.
This does not work --> sudo NETBOX_DELETE_LEGACY_DATA=1 upgrade.sh
Where to add -E ?

Thanks in advance, jr

**Reply by candlerb** [reply]
> This does not work --> sudo NETBOX_DELETE_LEGACY_DATA=1 upgrade.sh

You didn't show the error you got.

Almost certainly upgrade.sh is not in your path, so you need the `./` prefix

```
sudo NETBOX_DELETE_LEGACY_DATA=1 ./upgrade.sh
```

To answer your question, `-E` would go after `sudo`, but it shouldn't be needed here.

**Comment by jcresp21** [comment]
I am having the same Issue here, adding NETBOX_DELETE_LEGACY_DATA remvoes all Sites and Racks, any idea, I am trying to Update from 3.3.10 to 3.7 , Any Idea as which Database fields should be removed  ?

**Reply by candlerb** [reply]
> adding NETBOX_DELETE_LEGACY_DATA remvoes all Sites and Racks

It can't possibly do that: all devices have a not-nullable foreign key link to Site, so the only way it could do that is to delete all Devices as well.

Can you describe *exactly* what symptoms you are seeing?

**Reply by jcresp21** [reply]
Nothing out of the box, I have just  pushed a new docker compose version, adapted my env and conf files, point to my DB, and docker compose up, first run I did not add the NETBOX_DELETE_LEGACY_DATA , I got this error: 
avanetbox-netbox-1               |   File "/opt/netbox/netbox/circuits/migrations/0040_provider_remove_deprecated_fields.py", line 24, in check_legacy_data
avanetbox-netbox-1               |     raise DataError(
avanetbox-netbox-1               | django.db.utils.DataError: Unable to proceed with deleting contact fields from Provider model: Found 2 providers with legacy contact data. Please ensure all legacy provider contact data has been migrated to contact objects before proceeding. Or, set the NETBOX_DELETE_LEGACY_DATA environment variable to bypass this safeguard and delete all legacy provider contact data.
avanetbox-netbox-1 exited with code 1
After this I added NETBOX_DELETE_LEGACY_DATA=1 and it finished the upgrade process fine, I realized when I entered the UI that Sites and Racks weren't present, don't really know if something else, I am trying to remove manually now the old legacy Data before trying another update ( I went back to version 3.3.10 and restored db from backup)

**Reply by jcresp21** [reply]
I removed manually the legacy data and tried updating without the NETBOX_DELETE_LEGACY_DATA=1, everything went right and I haven't seen anything missing yet, I'm wondering if there was something else that happened while migrating the DB, I tried different times, so I don't have logs at hand. 
Thanks for the reply :-)
